{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5f8564dd-31ce-4dc6-baf6-a49c3dcc05cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %% import modules\n",
    "import geppy as gep\n",
    "from deap import creator, base, tools\n",
    "import numpy as np\n",
    "import random\n",
    "import operator \n",
    "import pickle\n",
    "from fractions import Fraction\n",
    "import scipy.io as scio\n",
    "import time\n",
    "import sympy as sp\n",
    "import copy\n",
    "import DHC_MGEP_loss as dm\n",
    "\n",
    "#For reproduction\n",
    "s = 0\n",
    "random.seed(s)\n",
    "np.random.seed(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "70abb123-bcab-442b-af39-481f19c27d7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %% define functions for tensor operators \n",
    "def tensor_add(a, b):\n",
    "    if isinstance(a, np.ndarray) and isinstance(b, np.ndarray) and a.shape == b.shape:\n",
    "        return a + b\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def tensor_sub(a, b):\n",
    "    if isinstance(a, np.ndarray) and isinstance(b, np.ndarray) and a.shape == b.shape:\n",
    "        return a - b\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def tensor_inner_product(a, b):\n",
    "    if isinstance(a, np.ndarray) and isinstance(b, np.ndarray) and a.shape == b.shape:\n",
    "        result = np.empty_like(a)\n",
    "        for i in range(a.shape[0]):\n",
    "            result[i] = np.matmul(a[i], b[i])\n",
    "        return result\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def p_(tensor):\n",
    "    '''\n",
    "    need to define a global variable ---- plasmid_library through function Y_pretend\n",
    "    '''\n",
    "    if isinstance(tensor, np.ndarray):\n",
    "        scalar = eval(plasmid_library.pop(0))\n",
    "        if isinstance(scalar, np.ndarray):\n",
    "            return scalar[:, None] * tensor\n",
    "        return tensor * scalar\n",
    "    else:\n",
    "        return False\n",
    "        \n",
    "# define a protected division to avoid dividing by zero\n",
    "def protected_div(x1, x2):\n",
    "    if isinstance(x2, np.ndarray):\n",
    "        abs_x2 = np.maximum(x2,-x2)\n",
    "        if (abs_x2 < 1e-10).any():\n",
    "            return 1\n",
    "        return x1 / x2\n",
    "    else:\n",
    "        if abs(x2) < 1e-10:\n",
    "            return 1\n",
    "        return x1 / x2\n",
    "\n",
    "def p_symbol(tensor):\n",
    "    '''\n",
    "    need to define a global variable ---- plasmid_library through function: my_compile\n",
    "    '''\n",
    "    scalar_expression = plasmid_library.pop(0)\n",
    "    return tensor * scalar_expression\n",
    "    \n",
    "def add(a,b):\n",
    "    return a + b\n",
    "\n",
    "def sub(a,b):\n",
    "    return a - b\n",
    "\n",
    "def mul(a,b):\n",
    "    return a * b\n",
    "\n",
    "symbolic_function_map = {\n",
    "        'tensor_add': operator.add,\n",
    "        'tensor_sub': operator.sub,\n",
    "        'tensor_inner_product': operator.mul,\n",
    "        'p_': p_symbol,\n",
    "        operator.add.__name__: operator.add,\n",
    "        operator.sub.__name__: operator.sub,\n",
    "        operator.mul.__name__: operator.mul,\n",
    "        'protected_div': operator.truediv,\n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "24234f6c-b26c-4e64-a771-b6e97c2e13a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %% generate data\n",
    "\n",
    "# target equation: tau_ij = -2*miu*S_ij + (2/3)*miu*D_kk*delta_ij + p*delta_ij\n",
    "#                         = -miu*(u_i__j+u_j__i) + (2/3)*miu*D_kk*delta_ij + p*delta_ij\n",
    "\n",
    "data = scio.loadmat('data/velocity_gradient_new.mat')\n",
    "u_x = data['u_x']\n",
    "u_y = data['u_y']\n",
    "v_x = data['v_x']\n",
    "v_y = data['v_y']\n",
    "\n",
    "target_data = np.loadtxt(f'data/TecGrid-Cavity_Kn0.01_new.dat', delimiter=' ', skiprows = 3)\n",
    "p_xx = target_data[:,7:8]\n",
    "p_xy = target_data[:,10:11]\n",
    "p_yx = p_xy\n",
    "p_yy = target_data[:,8:9]\n",
    "\n",
    "n_rho = target_data[:,2:3]\n",
    "k = 1.38065e-23\n",
    "T = target_data[:,6:7]\n",
    "p = n_rho * k * T\n",
    "\n",
    "x_data = data['x']  \n",
    "y_data = data['y'] \n",
    "\n",
    "x_range1 = (0.2, 0.8)\n",
    "y_range1 = (0.2, 0.8)\n",
    "x_indices = np.where((x_data >= x_range1[0]) & (x_data <= x_range1[1]))[0]\n",
    "y_indices = np.where((y_data >= y_range1[0]) & (y_data <= y_range1[1]))[0]\n",
    "\n",
    "indices0 = np.intersect1d(x_indices, y_indices)\n",
    "\n",
    "u_x = u_x[indices0]\n",
    "u_y = u_y[indices0]\n",
    "v_x = v_x[indices0]\n",
    "v_y = v_y[indices0]\n",
    "\n",
    "p_xx = p_xx[indices0]\n",
    "p_xy = p_xy[indices0]\n",
    "p_yx = p_yx[indices0]\n",
    "p_yy = p_yy[indices0]\n",
    "\n",
    "p = p[indices0]\n",
    "\n",
    "# subsample\n",
    "indices = np.random.choice(u_x.shape[0], 200, replace=False)\n",
    "u_x = u_x[indices]\n",
    "u_y = u_y[indices]\n",
    "v_x = v_x[indices]\n",
    "v_y = v_y[indices]\n",
    "\n",
    "p_xx = p_xx[indices]\n",
    "p_xy = p_xy[indices]\n",
    "p_yx = p_yx[indices]\n",
    "p_yy = p_yy[indices]\n",
    "\n",
    "p = p[indices]\n",
    "\n",
    "u_i__j = np.stack((np.stack((u_x, u_y), axis=1), np.stack((v_x, v_y), axis=1)), axis=2)\n",
    "u_j__i = np.stack((np.stack((u_x, v_x), axis=1), np.stack((u_y, v_y), axis=1)), axis=2)\n",
    "u_i__j = u_i__j.reshape((len(u_x), 2, 2))\n",
    "u_j__i = u_j__i.reshape((len(u_x), 2, 2))\n",
    "\n",
    "S_ij = 0.5 * (u_i__j + u_j__i)\n",
    "Omega_ij = 0.5 * (u_i__j - u_j__i)\n",
    "\n",
    "df_c = 1.399e-05 # Diffusion coefficient\n",
    "miu = 2.117e-5 # Viscosity coefficient\n",
    "\n",
    "D_kk = u_x + v_y\n",
    "det___D_ij = u_x * v_y - u_y * v_x\n",
    "\n",
    "shape = (len(u_x),1)\n",
    "ones = np.ones(shape, dtype = np.float64)\n",
    "zeros = np.zeros(shape, dtype = np.float64)\n",
    "delta_ij = np.stack((np.stack((ones, zeros), axis=1), np.stack((zeros, ones), axis=1)), axis=2)\n",
    "delta_ij = delta_ij.reshape((len(u_x), 2, 2))\n",
    "                            \n",
    "tau_ij = -2*miu*S_ij + p[:, None]*delta_ij + (2/3)*miu*D_kk[:, None]*delta_ij\n",
    "\n",
    "def add_noise(array, snr):\n",
    "    mean_power = np.mean(array**2)\n",
    "    noise_power = mean_power / (10**(snr / 10))\n",
    "    noise = np.random.normal(0, np.sqrt(noise_power), array.shape)\n",
    "    noisy_array = array + noise\n",
    "    return noisy_array\n",
    "\n",
    "def add_gaussian_noise(array, mean=0, std=1):\n",
    "    noise = np.random.normal(mean, std, array.shape)\n",
    "    return array + noise\n",
    "\n",
    "def add_uniform_noise(array, low=-1, high=1):\n",
    "    noise = np.random.uniform(low, high, array.shape)\n",
    "    return array + noise\n",
    "\n",
    "Y = tau_ij"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd9fe562-d945-423d-9fe6-56222ad002d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %% Assign number tags\n",
    "\n",
    "# Assign prime number tags to base dimensions\n",
    "L,M,T,I,Theta,N,J = 2,3,5,7,11,13,17\n",
    "\n",
    "# Derive the tags for dirived physical quantities according to their dimensions\n",
    "# Note that the tags are always in the form of fractions, instead of floats, which avoids introducing any truncation error. \n",
    "# Therefore, we use 'Fraction' function here.\n",
    "dict_of_dimension = {'S_ij':Fraction(1,T),\n",
    "                     'Omega_ij':Fraction(1,T),\n",
    "                     # 'u_i,j':Fraction(1,T),\n",
    "                     # 'u_j,i':Fraction(1,T),\n",
    "                     'delta_ij':Fraction(1),\n",
    "                     'D_kk':Fraction(1,T),\n",
    "                     'det___D_ij':Fraction(1,(T**2)),\n",
    "                     'miu':Fraction(M,L*T),\n",
    "                     'df_c':Fraction((L**2),T),\n",
    "                     'p':Fraction(M,L*((T)**(2)))}  \n",
    "\n",
    "# Assign number tags to taget variable\n",
    "target_dimension = Fraction(M,L*((T)**(2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "59d089b9-6b75-4ffe-bbac-d5ae95d6cb58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %% Creating the primitives set\n",
    "# Define the operators\n",
    "host_pset = gep.PrimitiveSet('Host', input_names=['S_ij','Omega_ij','delta_ij'])\n",
    "host_pset.add_function(tensor_add, 2)\n",
    "host_pset.add_function(tensor_sub, 2)\n",
    "host_pset.add_function(tensor_inner_product, 2)\n",
    "host_pset.add_function(p_, 1)\n",
    "\n",
    "\n",
    "plasmid_pset = gep.PrimitiveSet('Plasmid', input_names=['p','D_kk','det___D_ij'])\n",
    "plasmid_pset.add_symbol_terminal('df_c', df_c)\n",
    "plasmid_pset.add_symbol_terminal('miu', miu)\n",
    "plasmid_pset.add_function(operator.add, 2)\n",
    "plasmid_pset.add_function(operator.sub, 2)\n",
    "plasmid_pset.add_function(operator.mul, 2)\n",
    "plasmid_pset.add_function(protected_div, 2)\n",
    "plasmid_pset.add_rnc_terminal() # Add random numerical constants (RNC).\n",
    "# %% Create the individual and population\n",
    "\n",
    "# Define the indiviudal class, a subclass of gep.Chromosome\n",
    "creator.create(\"FitnessMax\", base.Fitness, weights=(-1,))  # weights=(-1,)/weights=(1,) means to minimize/maximize the objective (fitness).\n",
    "creator.create(\"Host_Individual\", gep.Chromosome, fitness=creator.FitnessMax, plasmid=[])\n",
    "creator.create(\"Plasmid_Individual\", gep.Chromosome) \n",
    "\n",
    "# Register the individual and population creation operations\n",
    "h = 5           # head length\n",
    "n_genes = 2      # number of genes in a chromosome\n",
    "r = 15            # length of the RNC array\n",
    "# enable_ls = True # whether to apply the linear scaling technique\n",
    "\n",
    "toolbox = gep.Toolbox()\n",
    "\n",
    "toolbox.register('host_gene_gen', gep.Gene, pset=host_pset, head_length=h)\n",
    "toolbox.register('host_individual', creator.Host_Individual, gene_gen=toolbox.host_gene_gen, n_genes=n_genes, linker=tensor_add)\n",
    "toolbox.register(\"host_population\", tools.initRepeat, list, toolbox.host_individual)\n",
    "\n",
    "def random_float(a, b):\n",
    "    # Generate floating-point numbers that retain five decimal places\n",
    "    num = random.uniform(a, b)\n",
    "    num = round(num, 5)\n",
    "    return num\n",
    "    \n",
    "toolbox.register('rnc_gen', random_float, a=-5.0, b=5.0)   # each RNC is random float within [-0.2, 0.2]\n",
    "toolbox.register('plasmid_gene_gen', gep.GeneDc, pset=plasmid_pset, head_length=h, rnc_gen=toolbox.rnc_gen, rnc_array_length=r)\n",
    "toolbox.register('plasmid_individual', creator.Plasmid_Individual, gene_gen=toolbox.plasmid_gene_gen, n_genes=n_genes, linker=operator.add)\n",
    "toolbox.register(\"plasmid_population\", dm.plasmid_generate, toolbox.plasmid_individual)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cebe35bb-93ed-49cd-a4c9-009dfcee89e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_expressed_plasmid(host_ind):\n",
    "    # Extract the plasmids to be expressed from the whole plasmids list in eval order and return the list\n",
    "    global plasmid_lis\n",
    "    global test_host_ind\n",
    "    test_host_ind = copy.deepcopy(host_ind)\n",
    "    plasmid_lis = host_ind.plasmid\n",
    "    et = dm.ExpressionTree.from_genotype(host_ind)\n",
    "\n",
    "    def preorderTraversal(tree):\n",
    "        stack = [tree.root]\n",
    "        result = []\n",
    "        while len(stack) != 0:\n",
    "            node = stack.pop()\n",
    "            if node.name == 'p_':\n",
    "                result.append(node.index)\n",
    "            if node.children is not None:\n",
    "                for child in node.children:\n",
    "                    if isinstance(child, dm.ExpressionTree):\n",
    "                        stack.append(child.root)\n",
    "                    else:\n",
    "                        stack.append(child)\n",
    "        return result[::-1]\n",
    "    \n",
    "    def renumber(q):\n",
    "        parsed = [s[1:-1].split('][') for s in q]\n",
    "        parsed = [[eval(i) for i in lis] for lis in parsed]\n",
    "        d = {}\n",
    "        for x, y in parsed:\n",
    "            if x not in d:\n",
    "                d[x] = []\n",
    "            d[x].append(y)\n",
    "        for x in d:\n",
    "            sorted_unique_values = sorted(d[x])\n",
    "            d[x] = {v: i for i, v in enumerate(sorted_unique_values)}\n",
    "        for i in parsed:\n",
    "            i[1] = d[i[0]][i[1]]\n",
    "        result = ['['+str(i)+']['+str(j)+']' for i, j in parsed]\n",
    "        return result\n",
    "        \n",
    "    index_lis = preorderTraversal(et)  \n",
    "    result = []\n",
    "    for index in renumber(index_lis):\n",
    "        if int(index[1:-1].split('][')[0]) >= len(plasmid_lis):\n",
    "            raise IndexError(\"list index out of range because\"+str(int(index[1:-1].split('][')[0]))+\">=\"+str(len(plasmid_lis)))\n",
    "\n",
    "        if int(index[1:-1].split('][')[1]) >= len(plasmid_lis[int(index[1:-1].split('][')[0])]):\n",
    "            print(index_lis)\n",
    "            print(renumber(index_lis))\n",
    "            print(index)\n",
    "            print(plasmid_lis)\n",
    "            raise IndexError(\"list index out of range because\"+str(int(index[1:-1].split('][')[1]))+\">=\"+str(len(plasmid_lis[int(index[1:-1].split('][')[0])])))\n",
    "        result.append(eval('plasmid_lis'+str(index)))\n",
    "\n",
    "    return result\n",
    "    \n",
    "def my_compile(Individual):\n",
    "    global plasmid_library\n",
    "    plasmid_library = []\n",
    "    plasmidind_lib = extract_expressed_plasmid(Individual)\n",
    "    for plasmid_ind in plasmidind_lib:\n",
    "        scalar_expression = gep.simplify(plasmid_ind, symbolic_function_map)\n",
    "        plasmid_library.append(scalar_expression)\n",
    "    tensor_expression = gep.simplify(Individual, symbolic_function_map)\n",
    "    return str(tensor_expression)\n",
    "\n",
    "def Y_pretend(Individual):\n",
    "    global plasmid_library\n",
    "    plasmid_library = [str(plasmid) for plasmid in extract_expressed_plasmid(Individual)]\n",
    "    host_str = str(Individual)\n",
    "    return eval(host_str)\n",
    "\n",
    "def dimensional_verification(individual, dict_of_dimension, target_dimension):\n",
    "    plasmid_library = [str(plasmid).replace('\\t','').replace('\\n','')\n",
    "                       .replace('add','my_add').replace('sub','my_sub')\n",
    "                       .replace('mul','my_mul').replace('truediv','my_truediv')\n",
    "                       .replace('protected_div','my_protected_div')\n",
    "                       for plasmid in extract_expressed_plasmid(individual)]\n",
    "    \n",
    "    def my_add(*args):\n",
    "        if any(isinstance(arg, bool) for arg in args):\n",
    "            return False\n",
    "        args = [arg for arg in args]\n",
    "        for i in range(len(args)):\n",
    "            if isinstance(args[i],int) or isinstance(args[i],float):\n",
    "                args[i] = Fraction(1)\n",
    "                \n",
    "        if all(arg == args[0] for arg in args):\n",
    "            return args[0]\n",
    "        return False\n",
    "        \n",
    "    def my_sub(a,b):\n",
    "        \n",
    "        if isinstance(a,bool) or isinstance(b,bool):\n",
    "            return False\n",
    "        if isinstance(a,int) or isinstance(a,float):\n",
    "            a = Fraction(1)\n",
    "        if isinstance(b,int) or isinstance(b,float):\n",
    "            b = Fraction(1)\n",
    "\n",
    "        if a == b:\n",
    "            return a\n",
    "        else:\n",
    "            return False\n",
    "\n",
    "    def my_mul(a,b):\n",
    "        \n",
    "        if isinstance(a,bool) or isinstance(b,bool):\n",
    "            return False\n",
    "        if isinstance(a,int) or isinstance(a,float):\n",
    "            a = Fraction(1)\n",
    "        if isinstance(b,int) or isinstance(b,float):\n",
    "            b = Fraction(1)\n",
    "            \n",
    "        return a * b\n",
    "\n",
    "    def my_truediv(a,b):\n",
    "        \n",
    "        if isinstance(a,bool) or isinstance(b,bool) or b == 0:\n",
    "            return False\n",
    "        if isinstance(a,int) or isinstance(a,float):\n",
    "            a = Fraction(1)\n",
    "        if isinstance(b,int) or isinstance(b,float):\n",
    "            b = Fraction(1)\n",
    "\n",
    "        return a / b\n",
    "\n",
    "    def my_protected_div(a,b):\n",
    "        \n",
    "        if isinstance(a,bool) or isinstance(b,bool) or b == 0:\n",
    "            return False\n",
    "        if isinstance(a,int) or isinstance(a,float):\n",
    "            a = Fraction(1)\n",
    "        if isinstance(b,int) or isinstance(b,float):\n",
    "            b = Fraction(1)\n",
    "\n",
    "        return a / b\n",
    "    \n",
    "    def my_p_(tensor):\n",
    "        if isinstance(tensor, bool):\n",
    "            return False\n",
    "        scalor = plasmid_library.pop(0)\n",
    "        scalor = eval(scalor, create_var)\n",
    "        return scalor * tensor\n",
    "\n",
    "        \n",
    "    create_var = locals()\n",
    "    create_var.update(dict_of_dimension)\n",
    "    namespace = {'my_add': my_add, 'my_sub': my_sub,'my_mul': my_mul,'my_truediv': my_truediv, 'my_protected_div': my_protected_div}\n",
    "    create_var.update(namespace)\n",
    "    \n",
    "    individual_expr = str(individual).replace('\\t','').replace('\\n','').replace('tensor_add','my_add').replace('tensor_sub','my_sub').replace('tensor_inner_product','my_mul').replace('p_','my_p_')\n",
    "    dimension_of_DDEq = eval(individual_expr)\n",
    "    if dimension_of_DDEq == target_dimension:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "        \n",
    "toolbox.register('compile',my_compile)\n",
    "toolbox.register('dimensional_verification',dimensional_verification)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "64a222ec-a494-436a-bb20-8dcb339a8261",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %% Define the loss function\n",
    "\n",
    "# Register the dimensional verification operation\n",
    "# Define the fitness for individuals that apply the linear scaling technique\n",
    "def evaluate(individual):\n",
    "    \"\"\"\n",
    "    First verify whether the individuals satisfy dimensional homogeneity.\n",
    "    If it is not dimensional homogeneous, we would identify it as an invalid individual and directly assign a significant error to it.\n",
    "    Otherwise, we evaluate its loss\n",
    "    \"\"\"\n",
    "    validity = toolbox.dimensional_verification(individual, dict_of_dimension, target_dimension)\n",
    "    if not validity:\n",
    "        return 1000,\n",
    "    else:\n",
    "        Yp = Y_pretend(individual)\n",
    "        if isinstance(Yp, np.ndarray):\n",
    "            sum = 0\n",
    "            for i in range(Y.shape[1]):\n",
    "                for j in range(Y.shape[2]):\n",
    "                    yp = Yp[:,i,j].reshape(-1,1)\n",
    "                    y = Y[:,i,j].reshape(-1,1)\n",
    "                    l2_norm_dealta_y =  np.sqrt(np.sum((yp-y)**2))\n",
    "                    l2_norm_y = np.sqrt(np.sum((y)**2))\n",
    "                    if l2_norm_y == 0:\n",
    "                        raise ValueError(\"Divided by 0!\")\n",
    "                    sum += l2_norm_dealta_y/l2_norm_y\n",
    "            fitness = sum/(Y.shape[1]*Y.shape[2])\n",
    "            return fitness,\n",
    "            \n",
    "        else:\n",
    "            raise ValueError(\"type of Yp must be np.ndarray.\")\n",
    "\n",
    "# todo: ls-evaluate\n",
    "toolbox.register('evaluate', evaluate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "35743678-d799-4340-8963-e0e0ea60e894",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %% Register genetic operators\n",
    "toolbox.register('select', tools.selTournament, tournsize=200) # Selection operator\n",
    "# 1. general operators for host population\n",
    "toolbox.register('mut_uniform', dm.mutate_uniform, host_pset = host_pset, \n",
    "                 func = toolbox.plasmid_individual, ind_pb=0.3, pb=1)\n",
    "toolbox.register('mut_invert', dm.invert, pb=0.2)\n",
    "toolbox.register('mut_is_transpose', dm.is_transpose, pb=0.2)\n",
    "toolbox.register('mut_ris_transpose', dm.ris_transpose, pb=0.2)\n",
    "toolbox.register('mut_gene_transpose', dm.gene_transpose, pb=0.2)\n",
    "toolbox.register('cx_1p', dm.crossover_one_point, pb=0.4)\n",
    "toolbox.register('cx_2p', dm.crossover_two_point, pb=0.4)\n",
    "toolbox.register('cx_gene', dm.crossover_gene, pb=0.2)\n",
    "# 2. general operators for plasmid population\n",
    "toolbox.register('mut_uniform_plasmid', gep.mutate_uniform, pset = plasmid_pset, ind_pb=0.05, pb=1)\n",
    "toolbox.register('mut_invert_plasmid', gep.invert, pb=0.1)\n",
    "toolbox.register('mut_is_transpose_plasmid', gep.is_transpose, pb=0.1)\n",
    "toolbox.register('mut_ris_transpose_plasmid', gep.ris_transpose, pb=0.1)\n",
    "toolbox.register('mut_gene_transpose_plasmid', gep.gene_transpose, pb=0.1)\n",
    "# 3. Dc-specific operators\n",
    "toolbox.register('mut_dc_plasmid', gep.mutate_uniform_dc, ind_pb=0.05, pb=1)\n",
    "toolbox.register('mut_invert_dc_plasmid', gep.invert_dc, pb=0.1)\n",
    "toolbox.register('mut_transpose_dc_plasmid', gep.transpose_dc, pb=0.1)\n",
    "toolbox.register('mut_rnc_array_dc_plasmid', gep.mutate_rnc_array_dc, rnc_gen=toolbox.rnc_gen, ind_pb='0.5p')\n",
    "toolbox.pbs['mut_rnc_array_dc_plasmid'] = 1 \n",
    "\n",
    "# %% Statistics to be inspected\n",
    "stats = tools.Statistics(key=lambda ind: ind.fitness.values[0])\n",
    "stats.register(\"avg\", np.mean)\n",
    "# stats.register(\"std\", np.std)\n",
    "stats.register(\"min\", np.min)\n",
    "stats.register(\"max\", np.max)\n",
    "\n",
    "# %% Launch evolution\n",
    "\n",
    "# Define size of population and number of generations\n",
    "n_pop = 1600             # Number of individuals in a host_population\n",
    "n_gen = 60000           # Maximum Generation\n",
    "tol = 1e-7               # Threshold to terminate the evolution\n",
    "output_type = 'Constitutive_equation_of_Newtonian_fluid_cavity_clean'     # Name of the problem\n",
    "isRestart = False        # 'True'/'False' for initializing the population with given/random individuals.\n",
    "pkl_filepath = None      # Path of .pkl file\n",
    "\n",
    "# If isRestart is 'True', read the given .pkl file to load the individuals as the first generation population.\n",
    "# If isRestart is 'False', initialize the first generation population with random individuals.\n",
    "if isRestart:\n",
    "    with open(pkl_filepath,'rb') as file:\n",
    "        host_pop  = pickle.loads(file.read())\n",
    "else:\n",
    "    host_pop = toolbox.host_population(n=n_pop) \n",
    "    plasmid_pop = toolbox.plasmid_population(host_pop)\n",
    "    for ind_host, ind_plasmid in zip(host_pop, plasmid_pop):\n",
    "        ind_host.plasmid = ind_plasmid \n",
    "\n",
    "# Only record the best three individuals ever found in all generations\n",
    "champs = 3 \n",
    "hof = tools.HallOfFame(champs)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41e44b5d-0e81-4215-9637-2e7677d259ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Evolve\n",
    "start_time = time.time()\n",
    "pop, log = dm.gep_simple(host_pop, plasmid_pop, toolbox, n_generations=n_gen, n_elites=1,\n",
    "                          stats=stats, hall_of_fame=hof, verbose=True,tolerance = tol,GEP_type = output_type)\n",
    "print(time.time() - start_time)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
